GoogleIoTCore.agent.nut

```diff
@@ -0,0 +1,1004 @@
+// MIT License
+//
+// Copyright 2018 Electric Imp
+//
+// SPDX-License-Identifier: MIT
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
+// EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+
+// GoogleIoTCore is an Electric Imp agent-side library which allows your agent code to work with Google IoT Core.
+
+const GOOGLE_IOT_CORE_ERROR_NOT_CONNECTED           = 1000;
+const GOOGLE_IOT_CORE_ERROR_ALREADY_CONNECTED       = 1001;
+const GOOGLE_IOT_CORE_ERROR_OP_NOT_ALLOWED_NOW      = 1002;
+const GOOGLE_IOT_CORE_ERROR_TOKEN_REFRESHING        = 1003;
+const GOOGLE_IOT_CORE_ERROR_ALREADY_REGISTERED      = 1004;
+const GOOGLE_IOT_CORE_ERROR_GENERAL                 = 1010;
+
+class GoogleIoTCore {
+
+    static VERSION = "1.0.0";
+
+    _debugEnabled   = false;
+
+    _projectId      = null;
+    _cloudRegion    = null;
+    _registryId     = null;
+    _deviceId       = null;
+
+    // Metafunction to return class name when typeof <instance> is run
+    function _typeof() {
+        return "GoogleIoTCore";
+    }
+
+    // Information level logger
+    function _log(txt) {
+        if (_debugEnabled) {
+            server.log("[" + (typeof this) + "] " + txt);
+        }
+    }
+
+    // Error level logger
+    function _logError(txt) {
+        if (_debugEnabled) {
```

Comment:
`_err` should log messages unconditionally and thus be used only for abnormal situations (which never happens under normal conditions). everywhere else it should be regular `_log` call.